content/algorithms/Isomorphism/isomorphism.md

```diff
@@ -0,0 +1,221 @@
+In fact, isomorphism is part of the problems known as NP. This means that we don't know any algorithm that runs in polynomial time.
+
++++
+
+### Applications (TODO: Should I move this to the end?)
+
+- Verification of equivalence of different representations of the design of an electronic circuit and communication networks. 
+- Image recognition.
+- Identification of chemical compounds and proteins.
+- Algorithms for fingerprint, facial and retina matching. 
+-  Clustering Algorithms on social networks.
+
++++
+
+## Isomorphism Algorithms
+**Naive Approach**
+
++++
+
+There are some initial properties that we can check to decide whether it's possible to have an isomorphism
+- G and H have the same amount of nodes and edges 
+- The degree sequence for G and H are the same
+
+These are necessary conditions but don't guarantee that 2 graphs are isomorphic. Let's see a small example:
+
+```{code-cell} ipython3
+plt.subplot(221)
+G = nx.cycle_graph(6)
+nx.draw_circular(G)
+plt.title("G", fontweight="bold")
+plt.subplot(222)
+H = nx.union(nx.cycle_graph(3), nx.cycle_graph(3), rename = ("s","d"))
+nx.draw_circular(H, node_color = "r")
+plt.title("H", fontweight="bold")
+plt.show()
+```
+
+```{code-cell} ipython3
+nx.faster_could_be_isomorphic(G, D)
+```
+
+These graphs are clearly not isomorphic but they have the same degree secuence. 
+
++++
+
+Another property we can check for is: 
+- Same number of cycles of a particular length, for example, triangles. 
+
+```{code-cell} ipython3
+nx.fast_could_be_isomorphic(G, D)
+```
+
+Checking this new property we can detect that the previous example graphs were not isomorphic. 
+
+We can go one step further and check the number of cliques. 
+
+```{code-cell} ipython3
+nx.could_be_isomorphic(G, D)
+```
+
+Again we can detect that G and D are not isomorphic. But these conditions are not enough to say that 2 graphs are isomorphic. Let's look at the following example: 
+
+```{code-cell} ipython3
+plt.subplot(221)
+G = nx.Graph([(1, 2), (2, 3), (3, 4), (4, 1), (2,4)])
+nx.draw_circular(G, with_labels = True, node_color="g")
+plt.title("G", fontweight="bold")
+
+plt.subplot(222)
+H = nx.Graph([(1, 2), (2, 3), (3, 4), (2,4), (3, 1)])
+nx.draw_circular(H, with_labels = True, node_color = "c")
+plt.title("H", fontweight="bold")
+plt.show()
+```
+
+```{code-cell} ipython3
+nx.could_be_isomorphic(G1, G2)
+```
+
+These graphs meet all the necessary conditions but they're not isomorphic.
+
++++
+
+### Some classes of graphs with solution in polynomial time
+- Trees
+- Planar graphs(In fact, planar graph isomorphism is O(log(n)))
+- Interval graphs
+- Permutation graphs
+- Circulant graphs
+- Bounded-parameter graphs
+    - Graphs of bounded treewidth
+    - Graphs of bounded genus
+    - Graphs of bounded degree
+    - Graphs with bounded eigenvalue multiplicity
+    - k-Contractible graphs (a generalization of bounded degree and bounded genus)
+
```

Comment:
Maybe we could right an example of why these cases can be solved in polynomial time?